WordPress/gutenberg
# Fix navigation link ui close focus management

Fixes https://github.com/WordPress/gutenberg/issues/58820


## What?

Focus when closing the link ui has not been managed. This PR fixes focus loss from:
- New link (+ appender) -> Escape (doesn't create link) -> Focuses closest sibling (adjacent or parent or navigation block)
- New link (+ appender) -> Create link -> Focuses the newly created navigation item
- Edit link ( primary + k) -> Escape -> Focuses navigation link item 
- Editing link from toolbar -> Focus sent back to link toolbar item
- Empty link click or primary + k -> Focuses navigation link item 
- All the same scenarios for submenus as well
- Bonus: Fixes an issue where having a url-like menu item will always be selected when the popover closes, instead of only on create

It also adds an extensive e2e test that checks for focus management and being able to create, use the link control, and delete navigation items and submenu items.

## Why?

Accessibility, focus management. There were severe focus loss issues across the navigation block, especially when using the link control popover.

## How?

Adds openedBy state to navigation-link and submenu components and handles the various situations.

## Testing Instructions for Keyboard





Check for focus loss when interacting with the navigation block link ui only using your keyboard
- New link (+ appender) -> Escape (doesn't create link) -> Focuses closest sibling (adjacent or parent or navigation block)
- New link (+ appender) -> Create link -> Focuses the newly created navigation item
- Edit link ( primary + k) -> Escape -> Focuses navigation link item 
- Editing link from toolbar -> Focus sent back to link toolbar item
- Empty link click (Add label) or primary + k -> Focuses navigation link item 
- All the same scenarios for submenu items as well as submenu parents

For the navigation label selection issue:
- Create a link like "https//example.com"
- example.com label should be selected when closing the popover using Escape
- deselect the text
- Open the link popover again 
- Escape to close it
- example.com should not be selected

## Screenshots or screencast 
This video shows all the steps in the "navigation manages focus for creating, editing, and deleting items" test in slowMo mode:

https://github.com/WordPress/gutenberg/assets/967608/ec52f469-1f23-4a7e-8535-e7dcb026eee7
